### PR TITLE
Add metrics to capture the match type for SCP studies (SCP-4155)

### PR DESCRIPTION
**Pull request by ehanna4**
For our OKR on adding author info to studies adding this metric will enable us to more easily see how many studies match the search based on term, phrase or author match. 

This PR will only add tracking for SCP studies (aka not Azul studies), since the OKR is around author search which isn't supported for Azul at this time.

There is definitely potential follow up work to add Azul match data and more study result info. However getting this bit out for OKR tracking seemed most important so I am planning to file a follow on ticket for that other work as long as there is no strong opposition.

To test:
- Pull this branch and run it locally
- Search from the home page using different terms and/or phrases
- Check out the network tab for events titles 'search' and check out the property scpStudiesMatchData
- Or check out mixpanel dev for the same

Screen Shot 2022-03-22 at 3 33 04 PM

### CODE REVIEW

**Review by devonbush — CHANGES_REQUESTED** [review]
Code looks good.  Please update 'should return search results using keywords' in search_controller_test.rb to test your new metrics object.

Also, it looks like if a combined keyword+facet search is done, this will return the # and types of keyword matches, regardless of whether they were filtered out by the facet results.  Is that intended? (this is probably a rare enough case that it doesn't really matter)

**Review by eweitz — CHANGES_REQUESTED** [review]
Good stuff!  I hope this is the first in a string of analytics enhancements for search.

I note one blocker: Mixpanel log properties should be flattened per SCP convention.  Even if nested properties are searchable (I suspect they aren't, or at least not nearly as easily), being consistent in format with other log properties is important for analyst UX.

Here's how nested properties look in Mixpanel -- not glanceable, and likely problematic for easy queryability:

Nested_property_in_Mixpanel__SCP_2022-03-23


I also suggest several non-blocking refinements on log property nomenclature ("matches" -> "results") and segmentation (e.g. `"scpStudiesMatchData: {'numMatchesByAuthor': 2}"` -> `"numResults:scp:author": 2`).

**Comment by ehanna4** [comment]
> Also, it looks like if a combined keyword+facet search is done, this will return the # and types of keyword matches, regardless of whether they were filtered out by the facet results. Is that intended? (this is probably a rare enough case that it doesn't really matter)

Is a combined keyword+facet search the inferred search? If so this will set the `matchData` to an empty {} so it will report nothing rather than incorrect data

**Comment by devonbush** [comment]
I should have been clearer -- by a combined keyword + facet search I meant if someone searches for the facet `Species: rattus norvegicus` and the text 'Smith', that will return no results to the user, since there are no studies that have both that facet and the text 'Smith'.  However, the text match "Smith" will find two studies.  I believe your change means that we will record the search as having two text matches, even though it returned none to the user.  I *think* that's desirable or at least "good enough for now", but just wanted to clarify

File: app/controllers/api/v1/search_controller.rb [hunk 1]
```diff
@@ -515,22 +522,59 @@ def self.escape_terms_for_regex(term_list:)
       def self.generate_mongo_query_by_context(terms:, base_studies:, accessions:, query_context:)
         case query_context
         when :keyword
+          puts('in keyword')
```
Comment (eweitz) [inline app/controllers/api/v1/search_controller.rb:525]:
```suggestion
```

**Review by eweitz — APPROVED** [review]
Thanks for addressing that issue, and incorporating the suggestions!

**Comment by ehanna4** [comment]
@devonbush  oh okay I think actually it will report nothing for match data if the facet filter is used, regardless if there's a match or not

only "pure" text searches have match data

Screen Shot 2022-03-24 at 9 44 36 AM
Screen Shot 2022-03-24 at 9 43 31 AM
Screen Shot 2022-03-24 at 9 44 07 AM

**Comment by devonbush** [comment]
thanks for clarifying!

**Review by devonbush — APPROVED** [review]

**Pull request merged**